src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -421,82 +423,185 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
                 return ScheduleSyncReadScatterAtOffsetAsync(handle, buffers, fileOffset, cancellationToken);
             }
 
-            if (CanUseScatterGatherWindowsAPIs(handle))
+            switch (buffers.Count)
             {
-                long totalBytes = 0;
-                int buffersCount = buffers.Count;
-                for (int i = 0; i < buffersCount; i++)
+                case 0:
+                    return CastValueTask(ReadAtOffsetAsync(handle, Memory<byte>.Empty, fileOffset, cancellationToken));
+                case 1:
+                    return CastValueTask(ReadAtOffsetAsync(handle, buffers[0], fileOffset, cancellationToken));
+            }
+
+            if (CanUseScatterGatherWindowsAPIs(handle)
+                && TryPrepareScatterGatherBuffers(buffers, default(MemoryHandler), out MemoryHandle[]? pinnedBuffers, out int totalBytes))
+            {
+                try
                 {
-                    totalBytes += buffers[i].Length;
+                    return ReadScatterAtOffsetSingleSyscallAsync(handle, pinnedBuffers, fileOffset, totalBytes,
+                        cancellationToken);
                 }
-
-                if (totalBytes <= int.MaxValue) // the ReadFileScatter API uses int, not long
+                catch
                 {
-                    return ReadScatterAtOffsetSingleSyscallAsync(handle, buffers, fileOffset, (int)totalBytes, cancellationToken);
+                    foreach (MemoryHandle memoryHandle in pinnedBuffers)
+                    {
+                        memoryHandle.Dispose();
+                    }
+
+                    throw;
                 }
             }
 
             return ReadScatterAtOffsetMultipleSyscallsAsync(handle, buffers, fileOffset, cancellationToken);
+
+            static async ValueTask<long> CastValueTask(ValueTask<int> task) =>
+                // we have to await it because we can't cast a VT<int> to VT<long>
+                await task.ConfigureAwait(false);
+        }
+
+        // Abstracts away the type signature incompatibility between Memory and ReadOnlyMemory.
+        private interface IMemoryHandler<T>
```

Comment:
That would be a good opportunity to use `abstract static` interface methods. Do I have to do anything else before I use them?

Reply:
`abstract static` interfaces are not really finished. For example, they are not handled by crossgen yet. I do not think we are ready to start using them in the BCL yet.

Reply:
I see. I added a TODO to remind us.